File: JavaScript/Medium/LongestRepeatingCharacterReplacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
 var characterReplacement = function(s, k) {
    
  let map = {};
  let max = 0; 
  let left = 0;

  for (let right = 0; right < s.length; right++) {
    map[s.charAt(right)] ? map[s.charAt(right)]++ : map[s.charAt(right)] = 1;
    let count = (right - left + 1) - getCountOfMostFrequentChar(map);
    while (count > k) {
      map[s.charAt(left)]--;
      left++;
      count = (right - left + 1) - getCountOfMostFrequentChar(map);
    }
    max = Math.max(max, right - left + 1);
  }
  
  return max; 
  
};

let getCountOfMostFrequentChar = function (map) {
  let max = 0;
  for (let key in map) {
    max = Math.max(max, map[key]);
  }
  return max;
}

/*

  Two pointers, sliding window, track (count of most frequent character - size of window)

  Time Complexity: O(n)
  Space Complexity: O(1) (at most 26 characters get mapped)

  Where n is the length of the string

*/



/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {
    
  let map = {};
  let max = 0;
  let left = 0;
  let maxFreq = 0;

  for (let right = 0; right < s.length; right++) {
    map[s.charAt(right)] ? map[s.charAt(right)]++ : map[s.charAt(right)] = 1;
    maxFreq = Math.max(maxFreq, map[s.charAt(right)]);
    let count = (right - left + 1) - maxFreq;
    while (count > k) {
      map[s.charAt(left)]--;
      left++;
      count = (right - left + 1) - maxFreq;
    }
    max = Math.max(max, right - left + 1);
  }
  
  return max; 
  
};

/*

  Optimized from O(n*26)
  * When the left pointer gets moved forward the maxFreq doesn't get updated
  * But checking against the wrong maxFreq doesn't change the max
  * The max only updates when a new maxFreq is found

  Time: O(n)
  Space: O(1)

*/

File: C#/Medium/LongestRepeatingCharacterReplacement.cs
public class Solution
{
    public int CharacterReplacement(string s, int k)
    {
        var charFrequency = new Dictionary<char, int>();
        int left = 0;
        int max = 0;
        for (int right = 0; right < s.Length; right++)
        {
            char c = s[right];
            if (charFrequency.ContainsKey(c)) charFrequency[c] += 1;
            else charFrequency.Add(c, 1);

            int mostFreqChar = GetMostFrequentChar(charFrequency);
            int len = right - left + 1;

            while (len - mostFreqChar > k && left < right)
            {
                charFrequency[s[left]] -= 1;
                left += 1;
                mostFreqChar = GetMostFrequentChar(charFrequency);
                len = right - left + 1;
            }

            max = Math.Max(max, len);
        }

        return max;
    }

    private int GetMostFrequentChar(Dictionary<char, int> charFrequency)
    {
        int max = 0;

        foreach (var pair in charFrequency)
        {
            max = Math.Max(max, charFrequency[pair.Key]);
        }

        return max;
    }
}

/*

  Two pointers, sliding window, track (count of most frequent character - size of window)

  Time Complexity: O(n)
  Space Complexity: O(1) (at most 26 characters get mapped)

  Where n is the length of the string

*/



public class Solution2
{
    public int CharacterReplacement(string s, int k)
    {
        var count = new Dictionary<char, int>(); // char, count
        int longest = 0;
        int max = 0;
        for (int right = 0, left = 0; right < s.Length; right++)
        {
            char c = s[right];
            int len = right - left + 1;

            if (!count.ContainsKey(c)) count.Add(c, 0);
            count[c] += 1;

            max = Math.Max(max, count[c]);
            if (len - max > k)
            {
                count[s[left]] -= 1;
                left += 1;
            }

            longest = right - left + 1;
        }

        return longest; 
    }
}

/*
 
  Small optimization to avoid re-calculating the max freqency more often than is needed.

  Time: O(n)
  Space: O(1)

  ------------------------------------------------------------------------------------

  Optimized from O(n*26)

  Key insights: It's not necessary to check the character count map for the most frequent character.

  This works because the only way for the window size to grow past the max (longest) is for the max
  character count for a given character to grow to be larger than the last max.

*/
